cohttp-lwt/src/client.ml

```diff
@@ -12,33 +14,340 @@ module Make (IO : S.IO) (Net : S.Net with module IO = IO) = struct
+          Queue.iter
+            (fun {res_r; _} -> Lwt.wakeup_later_exn res_r e)
+            pipeline.in_flight;
+          Queue.clear pipeline.in_flight;
+          Lwt_condition.broadcast_exn pipeline.condition e;
+        end;
+        Lwt.return_unit
+
+    let rec reader pipeline =
+      let ic = match pipeline.conn with
+        | Connecting _ | Closed | Failed _ -> assert false
+        | Full (ic, _) | Half ic -> ic
+      in
+      Response.read ic >>= fun res ->
+      match res with
+      | `Ok res ->
+        (* don't take from queue yet, because body may still be in flight *)
+        let {req; res_r; _} = Queue.peek pipeline.in_flight in
+
+        (* A response header to a HEAD request is indistinguishable from a
+         * response header to a GET request. Therefore look at the request. *)
+        if
+          match Response.has_body res with
+          | _ when Request.meth req = `HEAD -> false
+          | `No -> false
+          | `Yes | `Unknown -> true
+        then begin
+          let stream =
+            Body.create_stream
+              Response.read_body_chunk
+              (Response.make_body_reader res ic)
+          in
+          (* finalise could run in a thread different from the lwt main thread.
+           * You may therefore not call into Lwt from a finaliser. *)
+          let closed = ref false in
+          Gc.finalise_last
+            begin fun () ->
+              if not !closed then
+                Log.warn (fun m ->
+                    m
+                      "Body not consumed, leaking stream! Refer to \
+                       https://github.com/mirage/ocaml-cohttp/issues/730 for \
+                       additional details")
+            end
+            stream;
+          Lwt.wakeup_later res_r (res, Body.of_stream stream);
+          Lwt_stream.closed stream >>= fun () ->
+          closed := true;
+          Queue.take pipeline.in_flight |> ignore;
+          Lwt_condition.broadcast pipeline.condition ();
+          reader pipeline
+        end
+        else begin
+          Queue.take pipeline.in_flight |> ignore;
+          Lwt_condition.broadcast pipeline.condition ();
+          Lwt.wakeup_later res_r (res, `Empty);
+          reader pipeline;
+        end
+      | `Eof ->
+        begin match pipeline.conn with
+        | Full _ | Half _ when Queue.is_empty pipeline.in_flight ->
+          pipeline.conn <- Closed
+        | Full (_, oc) ->
+          Net.close_out oc;
+          pipeline.conn <- Failed Retry
+        | Half _ ->
+          pipeline.conn <- Failed Retry
+        | Failed _ -> ()
+        | Connecting _ | Closed -> assert false
+        end;
+        finalise pipeline
+      | `Invalid reason ->
+        let e = Failure ("Cohttp_lwt failed to read response: " ^ reason) in
+        pipeline.conn <- Failed e;
+        finalise pipeline
+
+    (* in shutdown and process I rely on mutex waiters receiving the lock in the
```

Comment:
Is there risk of this changing in the future?

Reply:
I suspect it won't change, but that's why I added the comment after all.

Reply:
Do you reckon we have a way of doing it without relying on an implementation detail?

Reply:
Yes, by making the FIFO `Queue` explicit and not relying on `Lwt_mutex` FiFo behaviour.
Now that I'm thinking about it, there indeed is a problem with the current implementation in that the length of the waiting requests ignores requests in the `Lwt_mutex` queue.

Reply:
Ok, if this is the case, then I think we should make it implementation independent and have an explicit queue

Reply:
done.